### ISSUE TITLE
Connection interruptions during Kubernetes deployment scaledowns on Calico 3.16.X to latest

### ISSUE THREAD

**Issue by kashook** [projectcalico/calico#4762]
## Your Environment

We run Calico on Amazon EKS. We use the [amazon-vpc-cni-k8s plugin](https://github.com/aws/amazon-vpc-cni-k8s) with the default configuration that is in place whenever a new EKS control plane is created, and we install Calico by following the instructions [here](https://docs.aws.amazon.com/eks/latest/userguide/calico.html). Our worker nodes are based on Amazon's [EKS optimized Amazon Linux AMI](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html).

Our current cluster is running Kubernetes 1.19.8, `amazon-vpc-cni-k8s` plugin version v1.7.5, and Calico version 3.15.1. Calico was installed from [this manifest](https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.7.5/config/v1.7/calico.yaml). (At the time this was installed, it was the latest version recommended by the documentation linked above. This version does _not_ use the Tigera operator).

## Current Behavior

In a test cluster, we have tested upgrading to the current Calico version recommended by the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/calico.html), which at the time this issue was logged installs version 1.13.2 of the Tigera Operator, which installs version 3.17.1 of Calico. Whenever we upgrade critical components such as Calico, we flood a test application in our cluster with lots of http requests and ensure that no requests are lost to ensure that our upgrades don't cause even a brief outage. Our test application is coded to shut down properly when it receives a SIGTERM, and the pod in which our app is running has a prestop hook that sleeps for several seconds, which is a widely documented method for ensuring that your pod stays up and can handle any last requests that it may receive while Kubernetes is removing the pod from the list of available endpoints during the scale down of a deployment. (One of many examples describing this can be found [here](https://learnk8s.io/graceful-shutdown)).

We have noticed with Calico version 3.16.X, all the way to the current latest release of 3.19.1, whenever a Kubernetes deployment scales down to fewer replicas, we have requests dropped around the time that the old replicas disappear from the cluster.  (I say 3.16.X, because the first 3.16 version I was able to get to run in my test cluster without crashing was 3.16.3.  I suspect the problem was introduced in 3.16.0). The actual upgrade of Calico appears to cause no issues. However, after the new Calico is running, scaling down pods in a deployment causes intermittent hanging and failed http requests to our application. If I downgrade Calico back to a 3.15.X version, the problem goes away. I also tested whether upgrading to the latest version of the `amazon-vpc-cni-k8s` plugin makes any difference, but it does not (nor do I know if it's actually relevant). We use horizontal pod autoscaling and have replicas scaling up and down all the time, so the fact that scaling down replicas can cause connectivity problems means we can't upgrade past our current Calico version.

## Expected Behavior

When scaling down replicas in a deployment while flooding a service with requests, connections should never be interrupted.

## Steps to Reproduce

Note that to flood our test service with requests, I used both `curl` and `httping`.  The `httping` utility is available to the package management systems for at least [Ubuntu](https://packages.ubuntu.com/bionic/httping) and [Arch Linux](https://archlinux.org/packages/community/x86_64/httping/). The `httping` utilty has a `-f` flag to flood a URL with requests. I ran curl in a loop in bash like this:

```
while :; do curl -w "%{http_code}  %{time_total}\n" -o /dev/null -sSL "http://whatever.domain"; done
```

Our test process was:

1. Create a pod in the Kubernetes cluster that has `curl` and/or `httping` to use as a client to ping other pods. I just used the latest ubuntu image from dockerhub, got a `bash` terminal in the pod with `kubectl exec -ti -n <namespace> <podname> -- bash`, and installed `curl` and `httping`.
1. Create a deployment that contains pods running an http server. Ideally the http server should shut down cleanly when the pod is terminated. At the very least it should also have a prestop hook that sleeps for at least several seconds to ensure 100% uptime as replicas scale down. A Kubernetes service object should be created in front of the pods to load balance traffic.
1. Ensure you are running Calico 3.15.x and using the `amazon-vpc-cni-k8s` plugin.  (Not sure if the `amazon-vpc-cni-k8s` plugin version is actually relevant but this matches our environment). Our current version is from https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.7.5/config/v1.7/calico.yaml, as instructed to install by an older version of https://docs.aws.amazon.com/eks/latest/userguide/calico.html
1. `kubectl exec` to get a shell in your client test pod in 3 or 4 separate terminals and start flooding the service fronting your pods with requests from each terminal. Let them run for a bit to be convinced no connection errors occur and everything is stable
1. Scale up the deployment by one or more replicas. Watch the httping/curl. There should be no failures
1. Scale the deployment down to its original number of replicas. Watch the httping/curl. There should be no failures.
1. Upgrade Calico, following https://docs.aws.amazon.com/eks/latest/userguide/calico.html
1. Note the httping/curl terminals should not have had any failures
1. Scale up the deployment by several replicas. Watch the httping/curl. There should be no failures
1. Scale the deployment down to its original number of replicas. Watch your httping/curl. This time, requests start intermittently hanging and dying. Connections will be intermittently dropped from now on.
1. Stop the httping/curl in all terminals and immediately restart. Connections still intermittently hang or die.
1. Stop the httping/curl in all terminals several minutes and restart. Connections are no longer dropped. If you now scale up and back down again, connections start intermittently hanging/dying again and continue until you stop pinging for several minutes
1. Stop httping/curl for several minutes, then run again to ensure connections have stopped dying
1. Downgrade Calico back to 3.15.x. (NOTE: you can't just deploy the old manifest to downgrade. See details on how we easily switched between Calico versions below).
1. Run httping/curl again. Scale deployments up and back down. Note no connections die.

I've done this repeatedly in our test cluster and it behaves this way every time. I tested on the latest released Calico version as well (3.19.1 at this time) and the same behavior occurs.

NOTE: Once Calico is upgraded to the version managed by the Tigera Operator, it can't be downgraded easily to the pre-operator version. The Tigera operator allows you to add the `unsupported.operator.tigera.io` annotation to objects it is managing that tells it not to undo any manual changes that are made. I was able to switch between Calico versions with a script that uses `kubectl patch` to add that annotation and change the image version on the appropriate daemonsets and deployments.

Example bash script to patch the Tigera operator managed Calico:

```
ver=v3.15.1

kubectl patch -n calico-system daemonset calico-node --patch "{\"apiVersion\":\"apps/v1\",\"kind\":\"DaemonSet\",\"metadata\":{\"annotations\":{\"unsupported.operator.tigera.io/ignore\":\"true\"},\"name\":\"calico-node\"},\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"name\":\"calico-node\",\"image\":\"docker.io/calico/node:${ver}\"}],\"initContainers\":[{\"name\":\"flexvol-driver\",\"image\":\"docker.io/calico/pod2daemon-flexvol:${ver}\"}]}}}}"

kubectl patch -n calico-system deployment calico-typha --patch "{\"apiVersion\":\"apps/v1\",\"kind\":\"Deployment\",\"metadata\":{\"annotations\":{\"unsupported.operator.tigera.io/ignore\":\"true\"},\"name\":\"calico-typha\"},\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"name\":\"calico-typha\",\"image\":\"docker.io/calico/typha:${ver}\"}]}}}}"

kubectl patch -n calico-system deployment calico-kube-controllers --patch "{\"apiVersion\":\"apps/v1\",\"kind\":\"Deployment\",\"metadata\":{\"annotations\":{\"unsupported.operator.tigera.io/ignore\":\"true\"},\"name\":\"calico-kube-controllers\"},\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"name\":\"calico-kube-controllers\",\"image\":\"docker.io/calico/kube-controllers:${ver}\"}]}}}}"

kubectl rollout status -n calico-system daemonset/calico-node
```

**Comment by stevehipwell** [comment]
This might be related to https://github.com/projectcalico/calico/issues/4518 which has been a known issue on any version higher than `v1.15.5` for quite a while. I have no idea why the AWS docs haven't been updated to reflect this, to the point that they updated their Helm chart to install an affected version. If this is the issue it's due to be fixed in `v1.20` which is due to drop by the end of the month.

**Comment by kashook** [comment]
I got a chance to do some testing with `v3.19.2` which also contains the fix, and the problem seems to be gone. Thanks for the info @stevehipwell!
